File: src/main/java/playground/PdfImageCustomPipeline.java
package playground;

import core.pipeline.impl.AccumulatorWithAlgBlock;
import core.pipeline.impl.AlgorithmHostBlock;
import features.dicomimage.util.DicomRepoRequestProcessors;
import features.reports.creators.RepoAccessorReportCreator;
import features.reports.pdf.DataTableReportToPdfElementConverter;
import features.reports.pdf.ImageSeriesToPdfElementConverter;
import features.reports.pdf.PdfElementsAccumulator;
import features.repository.LocalRepositoryCommander;
import features.repositoryaccessors.RepoFileSaver;
import features.repositoryaccessors.RepositoryAccessorBlock;
import features.repositoryaccessors.data.RepoRequest;

public class PdfImageCustomPipeline {

    public static void exec()
    {
        AlgorithmHostBlock dicomReader = new AlgorithmHostBlock<>(DicomRepoRequestProcessors.ReadDicomImageSeriesAlg);
        AlgorithmHostBlock tableReporter = new AlgorithmHostBlock<>(new DataTableReportToPdfElementConverter<>(new RepoAccessorReportCreator()));
        AlgorithmHostBlock imageReporter = new AlgorithmHostBlock<>(new ImageSeriesToPdfElementConverter());

        dicomReader.addListener(tableReporter);
        dicomReader.addListener(imageReporter);

        AccumulatorWithAlgBlock pdfBlock = new AccumulatorWithAlgBlock<>(new PdfElementsAccumulator("report"),2 );

        tableReporter.addListener(pdfBlock);
        imageReporter.addListener(pdfBlock);

        RepositoryAccessorBlock repoBlock = new RepositoryAccessorBlock<>(new LocalRepositoryCommander(),
                new RepoFileSaver(), "c:\\src\\reports");

        pdfBlock.addListener(repoBlock);

        RepoRequest init = new RepoRequest("c:\\src\\dicoms", new LocalRepositoryCommander());

        dicomReader.inputDataReady(null, init);

        System.out.println ("Pipeline finished");
    }
}

File: src/main/java/playground/PdfImageCustomPipeline.kt
package playground

import core.data.Data
import core.data.FileData
import core.data.medimage.ImageSeriesData
import core.pipeline.AccumulatorWithAlgBlock
import core.pipeline.AlgorithmHostBlock
import features.dicomimage.util.DicomRepoRequestProcessors
import core.pipeline.Pipeline
import features.reports.PdfElementData
import features.reports.creators.RepoAccessorReportCreator
import features.reports.pdf.PdfElementsAccumulator
import features.reports.pdf.asPdfElementData
import features.repository.LocalRepositoryCommander
import features.repositoryaccessors.RepoFileSaver
import features.repositoryaccessors.RepositoryAccessorBlock
import features.repositoryaccessors.data.RepoRequest

object PdfImageCustomPipeline {

    fun exec() {

        //Creating pipeline
        val pipe = Pipeline("pdf to dicom")

        //initializing blocks
        val seriesReaderBlock = AlgorithmHostBlock(
                DicomRepoRequestProcessors.ReadDicomImageSeriesAlg,
                pipelineKeeper = pipe)

        val tableReportBlock = AlgorithmHostBlock<ImageSeriesData, PdfElementData>(
                { x -> RepoAccessorReportCreator().execute(x).asPdfElementData() },
                "Table reporter", pipe)

        val imageReporter = AlgorithmHostBlock<ImageSeriesData, PdfElementData>(
                {x -> x.asPdfElementData()},
                "image reporter", pipe)

        val pdfBlock = AccumulatorWithAlgBlock(PdfElementsAccumulator(
                "report"),
                2,
                "Accumulator",
                pipe)

        val reportSaverBlock = RepositoryAccessorBlock<FileData, Data>(LocalRepositoryCommander(),
                RepoFileSaver(), "c:\\src\\reports")

        //making connections
        seriesReaderBlock.dataReady += tableReportBlock::inputReady
        seriesReaderBlock.dataReady += imageReporter::inputReady

        tableReportBlock.dataReady += pdfBlock::inputReady
        imageReporter.dataReady += pdfBlock::inputReady

        pdfBlock.dataReady += reportSaverBlock::inputReady

        //create initial data
        val init = RepoRequest("c:\\src\\dicoms", LocalRepositoryCommander())

        //print every new session record
        pipe.session.newRecord += { _, b -> println(b) }

        //initialize root block and run
        pipe.rootBlock = seriesReaderBlock
        pipe.run(init)
    }
}
